### PR TITLE
Market peer merger datatype spec

### PR DESCRIPTION
**Pull request by xiej2520**
# Description

This resolves https://github.com/sbu-416-24sp/orcanet-rust/issues/10

This is PR (2/3) for merging market_peer_merger_routes into main. I'm putting this up for review on this repo since I would have to make a PR against main in the class repo and that's too many files to review.

## Items in PR

  * FileInfo chunk hashing has been implemented, and the peernode has been reworked to use FileInfoHash instead of `String`ly typed File hashes throughout.
* Adding more type safety to peernode
  * `String` file paths have mostly been converted to use `PathBuf` and `&Path` where found.
  * Encoded user base64 strings have been converted to an `EncodedUser` newtype (it was difficult to figure out the difference between peer_id, id, name, EncodedUser before), with associated methods provided.
  * The peer node has been reworked to use a `LocalFileInfo` struct storing `FileInfo`, a path, and a price, with a `HashMap<FileInfoHash, LocalFileInfo>` storing all local data instead of 3 separate hashmaps.
* Further handlers added to the market by SeaHorses (still not fully functional, defer reviewing the market aside from compilation to the next PR)
  * The server code has been reorganized into `routes/{team name}`, with (at least dummy) endpoints set up for every team. Further reorganization and review of these can be pushed off to the next PR.

## Reviewing

* `market` and `routes` review can be deferred to the next PR, make changes and suggestions to the latest version of market_peer_merger_routes at https://github.com/daminals/orcanet-rust/pull/26
* Review the peer node implementation of FileInfo/FileInfoHash as well as EncodedUser
* Check that the token code still works. (I tested the local file download with the peer node using `cargo run --bin peer-node -F test_local_market`)
* Review the jobs and files code, I believe it should be up to date with the latest vesion.

### CODE REVIEW

**Comment by daminals** [comment]
Let's highlight the first pr on the main repo and see if we can get it merged, then we can make the pr on that repo

File: proto/src/lib.rs [hunk 1]
```diff
@@ -27,8 +41,23 @@ pub mod market {
...
+        pub fn as_str(&self) -> &str {
+            self.0.as_str()
         }
     }
 
+    impl fmt::Display for FileInfoHash {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "{}", self.0)
+        }
+    }
```
Comment (reaovyd) [inline proto/src/lib.rs:62]: This seems fine. Would just have to merge this in something like #25 to use this `FileInfoHash` than the library one we had.

File: peernode/src/peer-node/store/mod.rs [hunk 2]
```diff
@@ -1,19 +1,31 @@
-use crate::peer::MarketClient;
 use crate::producer;
-use crate::transfer::files;
-use crate::transfer::jobs::Jobs;
-use anyhow::Result;
+use crate::transfer::files::LocalFileInfo;
+use crate::transfer::{
+    files::{self, get_file_info},
+    jobs::{self, Jobs},
```
Comment (reaovyd) [inline peernode/src/peer-node/store/mod.rs]: Nitpick: remove jobs::self unused import (unless it's used later then feel free to ignore this)

**Review by reaovyd — APPROVED** [review]
Thanks for your work. Only just left one small nitpick comment. The other comment is just for us later to incorporate the `FileInfoHash` in the market lib in future PRs.

~~As for the file structure, I think we're going to now incorporate the file structure from @andrewctam in #26 ? Correct me if I'm wrong though.~~

Oh yeah nvm just read where it's going to be incorporated in a future PR alongside the protocol handler stuff with `check_holders` and `register_file` as well(?)

File: peernode/src/peer-node/store/mod.rs [hunk 3]
```diff
@@ -1,19 +1,31 @@
...
+use proto::market::{FileInfoHash, User};
 use serde::{Deserialize, Serialize};
-use std::{collections::HashMap, fs, path::PathBuf};
+use std::{
+    collections::HashMap,
+    fs,
+    path::{Path, PathBuf},
+};
 
 #[derive()]
```
Comment (reaovyd) [inline peernode/src/peer-node/store/mod.rs]: Actually just one more thing, but think we can just remove this empty `#[derive()]` (it's not from your changes, but wondering if you can just remove that)

**Pull request merged**